signalfx-java/src/main/java/com/signalfx/signalflow/ComputationHandler.java

```diff
@@ -0,0 +1,147 @@
+
+    /**
+     * Override to process job progress messages 
+     * @param message job progress
+     */
+    protected void onMessage(JobProgressMessage message) {}
+
+    /**
+     * Override to process data messages 
+     * @param message data
+     */
+    protected void onMessage(DataMessage message) {}
+
+    /**
+     * Override to process event messages
+     * @param message event
+     */
+    protected void onMessage(EventMessage message) {}
+
+    /**
+     * Override to process metadata messages 
+     * @param message metadata
+     */
+    protected void onMessage(MetadataMessage message) {}
+    
+    /**
+     * @return computation start time in milliseconds since midnight, January 1, 1970 UTC
+     */
+    public Long getStartTimeMs() {
+        return this.startTimeMs;
+    }
+
+    /**
+     * @return computation stop time in milliseconds since midnight, January 1, 1970 UTC
+     */
+    public Long getStopTimeMs() {
+        return this.stopTimeMs;
+    }
+    
+    @Override
+    public Computation call() throws ComputationAbortedException, ComputationFailedException,
+            SignalFlowException, IllegalStateException {
+
+        if (this.computation.getState() == State.STATE_COMPLETED) {
+            throw new IllegalStateException("computation is completed");
+        }
+        
+        this.startTimeMs = System.currentTimeMillis();
+        this.stopTimeMs = null;
+        
+        try {
+            // iterate computation messages and route to message handling methods
+            for (ChannelMessage message : computation) {
+
+                switch (message.getType()) {
+
+                case JOB_START:
+                    JobStartMessage jobStartMessage = (JobStartMessage) message;
+                    onMessage(jobStartMessage);
+                    break;
+
+                case JOB_PROGRESS:
+                    JobProgressMessage jobProgressMessage = (JobProgressMessage) message;
+                    onMessage(jobProgressMessage);
+                    break;
+
+                case DATA_MESSAGE:
+                    DataMessage dataMessage = (DataMessage) message;
+                    onMessage(dataMessage);
+                    break;
+
+                case EVENT_MESSAGE:
+                    EventMessage eventMessage = (EventMessage) message;
+                    onMessage(eventMessage);
+                    break;
+
+                case METADATA_MESSAGE:
+                    MetadataMessage metadataMessage = (MetadataMessage) message;
+                    onMessage(metadataMessage);
+                    break;
+
+                default:
+                    break;
+                }
+            }
+        } finally {
+            this.stopTimeMs = System.currentTimeMillis();
+            close();
+        }
+
+        return this.computation;
+    }
+
+    @Override
+    public void run() {
+        try {
+            call();
+        } catch (ComputationAbortedException ex) {
```

Comment:
If you're providing this class as a base for people to build on, you shouldn't impose on them that all errors are ignored and logged.